Получение контактов из ContactsContractProvider

polyblack opened the pull request:
- Запрос разрешения на чтение контактов в MainActivity
- Получение контактов и их данных в ContactRepository из поставщика контактов

polyblack committed:
- Запрос разрешения на чтение контактов в MainActivity
- Получение контактов и их данных в ContactRepository из поставщика контактов
_____________
- Запрос разрешений непосредственно во фрагмент, перед использованием
- shouldShowRequestPermissionRationale() с отображением объясняющего Toast
- Все переменные Contact иммутабельные
- Использование .use{} с курсорами

polyblack committed:
- Запрос разрешения на чтение контактов
- Обновлены методы получения контактов на обращение к поставщику контактов

huzzoid commented on app/src/main/java/com/polyblack/contactsapp/repository/ContactsRepository.kt:
```diff
@@ -1,19 +1,131 @@
+import android.provider.ContactsContract.CommonDataKinds.Email
 import com.polyblack.contactsapp.model.Contact
 
 class ContactsRepository {
```
Самое время сделать класс зависимым от контекста (передавать контекст в конструкторе) и методы пусть теперь работают с этим передаваемым контекстом.
т.е. это поможет избавиться от передачи контекста в качестве аргумента метода

huzzoid commented on app/src/main/java/com/polyblack/contactsapp/repository/ContactsRepository.kt:
```diff
@@ -1,19 +1,131 @@
+                                contactId
+                            )
+                        )
+                        if (numbers.isNotEmpty()) {
```
firstOrNull

huzzoid commented on app/src/main/java/com/polyblack/contactsapp/repository/ContactsRepository.kt:
```diff
@@ -1,19 +1,131 @@
+                        if (numbers.isNotEmpty()) {
+                            number1 = numbers[0]
+                            if (numbers.size > 1) {
+                                number2 = numbers[1]
```
getOrNull

huzzoid commented on app/src/main/java/com/polyblack/contactsapp/repository/ContactsRepository.kt:
```diff
@@ -1,19 +1,131 @@
+                    val emails = ArrayList<String>()
+                    getIfContactHasEmailsResult(context, contactId)?.let {
+                        emails.addAll(getEmails(it))
+                        if (emails.isNotEmpty()) {
```
firstOrNull && getOrNull

huzzoid commented on app/src/main/java/com/polyblack/contactsapp/repository/ContactsRepository.kt:
```diff
@@ -1,19 +1,131 @@
+                    val id: Int = (cursor.getString(
+                        cursor.getColumnIndex(ContactsContract.Contacts._ID)
+                    ).toInt()
+                            )
```
К чему  тут относится скобка?

huzzoid commented on app/src/main/java/com/polyblack/contactsapp/ui/fragments/contact_list/ContactListFragment.kt:
```diff
@@ -105,12 +106,69 @@ class ContactListFragment : Fragment(),
+            }
+            else -> {
+                activity?.let {
+                    ActivityCompat.requestPermissions(
```
У фрагмента нет такого метода?

huzzoid commented on app/src/main/java/com/polyblack/contactsapp/ui/fragments/contact_list/ContactListFragment.kt:
```diff
@@ -105,12 +106,69 @@ class ContactListFragment : Fragment(),
+        grantResults: IntArray
+    ) {
+        when (requestCode) {
+            1 -> {
```
Лучше вынести в константу

huzzoid commented on app/src/main/java/com/polyblack/contactsapp/ui/fragments/contact_list/ContactListFragment.kt:
```diff
@@ -105,12 +106,69 @@ class ContactListFragment : Fragment(),
+        }
+    }
+
+    private fun doOnPermissionResult() {
```
Имя метода и место его вызова выглядят странно. Предлагаю вывзвать реквест пермишнов отдельно так, как рекомендовано докой. А обойти такую ситуацию можно, например флагами. Если нужно подробнее - можно завтра созвониться

polyblack committed:
- Запрос разрешения на чтение контактов в MainActivity
- Получение контактов и их данных в ContactRepository из поставщика контактов
_____________
- Запрос разрешений непосредственно во фрагмент, перед использованием
- shouldShowRequestPermissionRationale() с отображением объясняющего Toast
- Все переменные Contact иммутабельные
- Использование .use{} с курсорами
_____________
- ContactRepository принимает в конструкторе context, в Сервисе теперь используется экземпляр репозитория, а не статические методы
- firstOrNull() и getOrNull() в методах получения номеров и emails в репозитории
- boolean поля для проверки привязки сервиса и разрешения, вызов метода запроса контактов в нескольких местах и выполнение в зависимости от значений этих полей
- вынос в константу permission request code
- обращение к requestPermission() не от активити, а от фрагмента
_____________
- Закрытие курсора emails через .use в методе getIfContactHasEmailsResult()
- Вынос инициализации переменной в конструктор в классе репозитория

huzzoid commented on app/src/main/java/com/polyblack/contactsapp/repository/ContactsRepository.kt:
```diff
@@ -1,19 +1,120 @@
 
-class ContactsRepository {
+class ContactsRepository constructor(_context: Context) {
+    private val context = _context
```
Вынеси в конструктор

huzzoid commented on app/src/main/java/com/polyblack/contactsapp/repository/ContactsRepository.kt:
```diff
@@ -121,8 +95,26 @@ class ContactsRepository {
+        return numbers
+    }
+
+    private fun getIfContactHasEmailsResult(contactId: Int): Cursor? =
```
А этот курсор потом закрывается?
